### ISSUE TITLE
UX: Add batch edit dialog to edit multiple pictures at once

### ISSUE THREAD

**Issue by jannewulf** [photoprism/photoprism#271]
First of all: I just discovered your project and really like it so far. I think it has great potential.

I wanted to edit the location of a group of photos, I think that is a common thing to do for DSLR or older photos. Right now I would have to click through all the photos that I selected and set the location for each. 

I think the more intuitive thing to happen if you select multiple photos and click the edit button is a batch editing mode. So you would set a information for all the selected images and the rest of the information would stay as it is.

This kind of relates to the discussion in #212, but is not restricted to label editing. So I opened a new Issue.

------------------------------------------------------------------------------------------------------------------------------------------------------------

**Acceptance Criteria:**
- [x] There MUST be a separate batch edit dialog
- [x] User MUST be able to change **at least** location, date and labels for a batch of photos
- [x] For fields that can have multiple values (e.g. keywords, labels) it MUST be possible to add a new value without overwriting all existing values
- [x] For fields that can have multiple values (e.g. keywords, labels) it MUST be possible to remove single values
- [x] Write unit tests
- [ ] Write acceptance test
- [ ] Update documentation

**Comment by lastzero** [comment]
You're right, we'll implement batch editing in a future release. Also changing locations should be more convenient. At the moment, you can only change lat and long manually.

**Comment by jannewulf** [comment]
I have not much experience with web development, but I would give it a try to implement this functionality. But there are some things that need to be discussed first.

I think we should discuss which fields should go into the edit dialog, in order to keep it clean. 
1. The most important are probably Date/Time and Location. The labels from object detection are also kinda important as I get lots of wrong classifications until now. 
   - Should go into batch edit dialog
2. There are also lots of information that are rarely modified, like description.
   - Not sure about this. I never used a description field, so I don't know if batch editing makes sense for this.
   - I saw that, in Google Photos, it is possible to edit the description in the extended info view (screenshot attached). 
3. And last but not least, there are fields that are not going to be edited in a software like this (Google Photos etc. replacement). These include Camera, Lens, Focal Length, ISO, Exposure, File names, ...
   - Should be moved to a read only extended info view.

I don't really get the Keywords field. Where do these come from? Not or not only from the object detection, I guess, as they are not the same as the labels. I was not able to find more info on that.

Screenshot of the Google Photos Extended Info view with the option to edit the description ("Beschreibung hinzufügen") there.
![](https://user-images.githubusercontent.com/7462597/79255691-0226bf80-7e87-11ea-9a02-e9385425c5d8.png)

**Comment by lastzero** [comment]
Think the difficult part here is user experience (UX). Updating multiple fields in a database is relatively simple. Since there is so much going on, we don't really have time to deeply think about this or make decisions. Starting a discussion based on wireframes / mockups might be a good idea. I'm personally not impressed by the UX of Google Photos, so my first try wouldn't be to copy their user interface.

Maybe something simple like an "[x] apply changes to all selected photos" checkbox at the bottom would do for the start? I don't want to judge if a user really needs to change the camera for a photo or not, so you could change what you want and just apply it to all currently selected photos.

Keywords come from various sources such as file & directory names, colors and the location. You can see them in the dialog to add more or change words that are wrong. Still subject to improvement.

For the location, we want to use Places to show where you've been in the past (a path through time) that then can be used to estimate photos locations. Editing every single photo seems way too complicated.

**Comment by kwm1800** [comment]
I mean, it does not have to be perfect from start.

Right now all I really need, for example, is an ability to select all photos so I can sort them out by assigning them into albums.

Photoprism already have several ways to sort out pictures such as Colors, Lenses, Cameras, Years, etc. What missing is not having to click each of 100s pictures to move them into specific albums.

**Comment by lastzero** [comment]
You can either do a long click or shift-click to select multiple photos at once. Select all doesn't make much sense with infinite scrolling, see discussion in #272. However, we're going to implement search filters ("smart albums").

**Comment by kwm1800** [comment]
> 
> 
> You can either do a long click or shift-click to select multiple photos at once. Select all doesn't make much sense with infinite scrolling, see discussion in #272. However, we're going to implement search filters ("smart albums").

I see. If that is the case, is it possible to assign pictures into a specific album during the import process? 

Currently my pictures are all organized in folders, and I import files by folders so this could make organizing files faster.

**Comment by lastzero** [comment]
> I see. If that is the case, is it possible to assign pictures into a specific album during the import process?

See "Feature Request: Frontend - assign Album at the time of upload" #246
 
> Currently my pictures are all organized in folders, and I import files by folders so this could make organizing files faster.

Yes, it will :) We're already working on it, give us a little bit of time to find a great solution that works for everyone. Data is there, we just need to evaluate it and build a simple enough UI.

**Comment by PiotrEsse** [comment]
Hello, reffering to UX, theres already a way to edit multiple files but need little correction. Now, You can select many files and choose to edit by clicking edit icon. But now editing data in edit window only apply changes to ONE file. If this would be corrected to apply changes to all selected files this would solve the issue. Of course this can be optimize in the future but we would have basic functionality quick.

**Comment by lastzero** [comment]
We will definitely implement batch editing once our first release is out of the door, probably part of a special release for our sponsors and contributors.

**Comment by JohanAR** [comment]
Suggestion for how UI for keyword batch editing could work:

1. Find all keywords from all selected photos
2. Display each keyword with count, e.g. "cat (5)" if there are 5 photos that have the "cat" keyword.
3. Each displayed keyword has it's own - (minus) button, and if this is pressed the keyword gets red background, crossed over, moved to a different area or something else to indicate that it will be removed.
4. Below the keyword list there is a text entry box and a + (plus) button to enter new keywords that should be added.
5. Changes are applied when you press Done on the dialog.

[similar to this](https://penguin-arts.com/wp-content/uploads/2018/04/bootstrap-tags-input.png)

**Comment by lucanello** [comment]
Is there any update to this feature? I would also love to see it implemented!

**Comment by graciousgrey** [comment]
We had no time to to work on in yet. We are still busy finishing our first release. Batch edit will be one of the first features after that release :)

**Comment by stellarpower** [comment]
> Hello, reffering to UX, theres already a way to edit multiple files but need little correction. Now, You can select many files and choose to edit by clicking edit icon. But now editing data in edit window only apply changes to ONE file. If this would be corrected to apply changes to all selected files this would solve the issue. Of course this can be optimize in the future but we would have basic functionality quick.

I agree with this, I've imported about 25k photos and having to edit every single one individually is likely going to mean I'll have to shelve this for 6 months until this feature exists in some form. There may be various UI options, but using the multiple-selection feature would be a simple way of doing this if the backend can support updating multiple entries in the database transaction well; having the results of the multiple-selection only apply to the first file is counter-intuitive, and it would get this feature off the ground whilst any future UX improvements could be considered in more detail.

**Comment by lastzero** [comment]
We're doing our best do do this quickly. Finding and fixing any indexer issue has priority at the moment, so that we can tag a stable release. Financial support makes a huge difference and enables us to spend more time with this project.

**Comment by ndebruin** [comment]
In the mean time, is there a way to manually batch edit metadata? Whether through SQL, or through editing sidecar .yml files? I attempted to remove labels via sed in the sidecar .yml files, but the changes were not shown.

**Comment by lucanello** [comment]
> In the mean time, is there a way to manually batch edit metadata? Whether through SQL, or through editing sidecar .yml files? I attempted to remove labels via sed in the sidecar .yml files, but the changes were not shown.

I switched to using Digikam for these purposes, regarding photo and album organization as well as metadata editing, Digikam is the go-to open source solution.

**Comment by Ramblurr** [comment]
Is there a way to use Digikam and photoprism together harmoniously?

**Comment by JamesSwift** [comment]
> Suggestion for how UI for keyword batch editing could work:
> similar to this

Yes, great idea!

If you've ever used MP3Tag to edit music metadata their interface (if clunky) works well for batch editing. Fields that are the same across all selected items just show their text. Fields that are blank are, well, blank. And fields whose content varies are labeled [keep] - if left along that field won't be modified, if altered all selected items get the new value.

[MP3Tag Example](https://community.mp3tag.de/uploads/default/467fd9c05a65c2329604c548d869a504fa893ef2)

This works great, but for fields that contain list, like the keywords field, @JohanAR's solution would be excellent.

**Comment by GraysonCAdams** [comment]
> In the mean time, is there a way to manually batch edit metadata? Whether through SQL, or through editing sidecar .yml files? I attempted to remove labels via sed in the sidecar .yml files, but the changes were not shown.

```UPDATE photos SET photo_lat = 26.443398, photo_lng = -82.111511 WHERE photo_title = "Location Name?"```

Do backups if you're going to do it this way though.

**Comment by inthreedee** [comment]
The API can also be used to script batch updates. See a (very inefficient) example here:
https://github.com/inthreedee/photoprism-add-description

**Comment by timdonovanuk** [comment]
+1. The current UX/UI is confusing around this. In a folder, I have the option to select files and then press Edit..but I'm not actually editing the selected files. I'm not even sure why the ability to select files is an option in Folder view because all of the available options are actions that take place on the folder not the selected files.

**Comment by lastzero** [comment]
Selecting and editing files is the same in all views, including folders:

- You can either click a picture title to start editing- the prev / next buttons in the edit dialog will then get you to the previous or next pictures in the results list / folder contents.
- If you select result items and use the context menu, the prev / next buttons in the edit dialog will then get you to the previous or next picture in your selection, so that you don't need to open and close the dialog all the time.

Makes sense to me?

**Comment by wodka** [comment]
For changes regarding the date it would also be nice to only set a "start" date for the first image - so then all other images are moved to the time after the first one.

Why? Sometimes the date of the camera that took the photo is off, then all the dates are in the right order, with the right offsets but just shifted to the invalid date.

Example:

Image Name | Date of File | Date should be
------------ | -------------|--------
01.jpg | 2001-01-01 12:30 | 2021-04-15 14:30
02.jpg | 2001-01-01 13:00 | 2021-04-15 15:00
etc

**Comment by Mike-Sundays** [comment]
Is there any update about this feature?

**Comment by thomaseizinger** [comment]
Checkout the roadmap for what is currently being worked on https://github.com/photoprism/photoprism/projects/5#column-11831487.

**Comment by spiezmaestro** [comment]
Would a batch edit of "unassigned faces" via a view of "unassigned faces" be part of this? So I can select all the relevant faces (from a search like "face:new year:2021 month:11")and all assign them to a person in one go? Currently going picture by picture is tedious. Or should I open a new issue for this?

**Comment by graciousgrey** [comment]
@spiezmaestro I guess the following issues will solve your issue: https://github.com/photoprism/photoprism/issues/1679#issuecomment-962585995 and https://github.com/photoprism/photoprism/issues/1556

**Comment by isleshocky77** [comment]
> Selecting and editing files is the same in all views, including folders:
> 
> * You can either click a picture title to start editing- the prev / next buttons in the edit dialog will then get you to the previous or next pictures in the results list / folder contents.
> * If you select result items and use the context menu, the prev / next buttons in the edit dialog will then get you to the previous or next picture in your selection, so that you don't need to open and close the dialog all the time.
> 
> Makes sense to me?

This makes sense; however, if you change data and then click next, you lose those changes unless you hit apply first. It would be nice to keep the changed data and then hitting apply would save all changes.

**Comment by isleshocky77** [comment]
> Would a batch edit of "unassigned faces" via a view of "unassigned faces" be part of this? So I can select all the relevant faces (from a search like "face:new year:2021 month:11")and all assign them to a person in one go? Currently going picture by picture is tedious. Or should I open a new issue for this?

One thing to mention is that the view for search "face:new" doesn't show "faces", it shows images containing faces. So this batch edit falls down if there is more than one face in the image. I would like to see a page which just shows `markers` which don't have a `subject` and allow batch editing of those.

**Comment by graciousgrey** [comment]
That's planned as well: https://github.com/photoprism/photoprism/issues/1556 & https://github.com/photoprism/photoprism/issues/1679

**Comment by MarshallAsch** [comment]
I was thinking of a possible implementation for this, since its a feature I would really like. 

I was looking through the source code for the photo edit dialog and it looks like when the form could be changed to send the update api call for all images in the `selection` array. The values can be filled into the form using the following general scheme:

- all images have the same value, then show the value in the field. 
- if no images have a value, then show the same no value prompt that is currently shown
- if there are multiple different values, show something like `[mixed values]` and no not update those fields when the api call is made unless that was replaced. 

To get the values I was originally thinking of just calling find over and over again then generating that merged list, but that seems like it will potentially require loading a lot of items in the client. Alternatively I was thinking of something like a 
`GET /api/photos/bulk?uids=<comma separated list of ids>` that will do the processing on the server.  

I also thought about potentially adding a batch update api endpoint that will take the fields and the uids of the photos to update, perhaps that would be better.

I have not done any go development before but I have worked on several other projects. Id like to try giving it a shot to implement this and would like to check that my approach makes sense to the project is isnt a waste of everyones time.

**Comment by IssueHuntBot** [comment]
[@oshw-rico](https://issuehunt.io/u/oshw-rico) has funded $33.00 to this issue.

---
- Submit pull request via [IssueHunt](https://issuehunt.io/repos/119160553/issues/271) to receive this reward.
- Want to contribute? Chip in to this issue via [IssueHunt](https://issuehunt.io/repos/119160553/issues/271).
- Checkout the [IssueHunt Issue Explorer](https://issuehunt.io/issues) to see more funded issues.
- Need help from developers? [Add your repository](https://issuehunt.io/r/new) on IssueHunt to raise funds.

**Comment by lastzero** [comment]
Please don't add funds to IssueHunt anymore! 🎁

While we like IssueHunt and are grateful for the donations we've received so far, it hasn't proven to be a sustainable funding option for us as we spend much of our time maintaining existing features and providing support.

If we don't have enough resources to provide support and bugfixes, we can't start working on new features.

**Comment by IssueHuntBot** [comment]
[@averymd](https://issuehunt.io/u/averymd) has funded $5.00 to this issue.

---
- Submit pull request via [IssueHunt](https://issuehunt.io/repos/119160553/issues/271) to receive this reward.
- Want to contribute? Chip in to this issue via [IssueHunt](https://issuehunt.io/repos/119160553/issues/271).
- Checkout the [IssueHunt Issue Explorer](https://issuehunt.io/issues) to see more funded issues.
- Need help from developers? [Add your repository](https://issuehunt.io/r/new) on IssueHunt to raise funds.

**Comment by dankasak** [comment]
> Please don't add funds to IssueHunt anymore! gift
> 
> While we like IssueHunt and are grateful for the donations we've received so far, it hasn't proven to be a sustainable funding option for us as we spend much of our time maintaining existing features and providing support.
> 
> If we don't have enough resources to provide support and bugfixes, we can't start working on new features.

This sounds contradictory - ie don't add funds because you don't have enough resources. Perhaps a more detailed explanation would help. Also, the impression I get is that this ( in my opinion essential ) feature won't be implemented due to a lack of resources. Is that correct?

More broadly, perhaps you should consider a commercial version to help fund development?

**Comment by lastzero** [comment]
👉 https://docs.photoprism.app/funding/

**Comment by ThatCoffeeGuy** [comment]
Will this ever be a thing? This is the first time I see people literally paying for a feature on github like this. I can also see it's not working at all. :(

**Comment by lastzero** [comment]
*I deleted my sarcastic comment as these thumbs down clickers really made me think. We understand that people unfamiliar with software development may not realize how much time and money it takes to implement such a major feature (while providing daily support to 10,000+ users, most of whom are not sponsors). It should not be an exception to the rule when people pay for software they use... that is, the work we do 7 days a week, which goes way beyond what most users would consider acceptable for themselves.*

**Comment by ThatCoffeeGuy** [comment]
Because I'm not a web developer. I was considering funding it another couple of dollars, but then I saw it is already funded by a number of people, so I asked if it's pointless at this point, since it doesn't seem to be working. This reaction is something I wasn't expecting. 

And now I found your comment starting "Please don't add funds to IssueHunt anymore!" which is my fault, I missed it. To my defense I can see people kept funding it after your comment...

Edit: I can see it's in "upcoming" in the roadmap, which is promising. Regardless of that reaction, I appreciate this product and your work on it.

**Comment by graciousgrey** [comment]
What helps the most speed up development is becoming a sponsor.  We offer multiple [funding options](https://docs.photoprism.app/funding/). 
You are right , Issuehunt is not a sustainable funding option for us, that's why we don't use it anymore. Maintaining a project this size and complexity is a lot more work than just adding new features.  

Batch edit is planned and will be implemented but given our [current funding](https://github.com/sponsors/photoprism?preview=true) and [team size](https://docs.photoprism.app/team/) we can't work faster than we [do already](https://docs.photoprism.app/release-notes/).  
We would love to have all the features on our to do list implemented and they will be implemented - it just takes time :)

We have a zero bug policy and do our best to help users when they need support or have other questions. Updating and improving the documentation takes up a large part of our time too.

This comes at a price, as we can't give exact deadlines for new features.

**Comment by ThatCoffeeGuy** [comment]
> _I deleted my sarcastic comment as these thumbs down clickers really made me think. We understand that people unfamiliar with software development may not realize how much time and money it takes to implement such a major feature (while providing daily support to 10,000+ users, most of whom are not sponsors). It should not be an exception to the rule when people pay for software they use... that is, the work we do 7 days a week, which goes way beyond what most users would consider acceptable for themselves._

"We understand that people unfamiliar with software development may not realize how much time and money it takes to implement such a major feature"

You are really missing my point. We / I are completely aware of how much work and time it takes. I just wasn't sure about this model where you pay to get a feature implemented, because I didn't think it would work. You said yourself it doesn't work, but I missed that comment, which I also mentioned above. 

But you can also try to understand my point of view when making that comment: Not knowing about your 7 days a week workload at all, seeing a well-funded feature that wasn't implemented since 2020. Of course I'd wonder if this model works at all. 

After [graciousgrey](https://github.com/graciousgrey)'s comment it was a different story. It is really concerning to read that you guys work 7 days a week. As you casually mention must of us wouldn't do it - you're right. It also doesn't seem like even +$500 a month would resolve the issues with the workload, as that is barely enough to hire someone full-time. 

Please try to relax and take care.

**Comment by lastzero** [comment]
**Maintaining this project day in and day out is tough. That's why I enjoy being sarcastic from time to time, like most IT professionals I've met in my 25+ years in the industry. It's the little bit of luxury we have that doesn't cost anything.**

Don't take it away and stop giving personal advice to people you don't know. It's no secret that the open source community has a funding and entitlement problem. We are fully aware of this and have a strategy. Note we do not have a dedicated marketing team and also do not have much time to think about every public comment.

Thank you! 🙏

**Comment by beefstew** [comment]
UX suggestion on the image selection mechanism:
Copy feature from Photos on iOS where user can brush finger across multiple images to select them. The general idea is to allow selection of photos using gestures that are fast and imprecise.

Currently, it is very cumbersome to select multiple images to apply a common tag; I must hover each image thumbnail and click selection circle.

**Comment by graciousgrey** [comment]
@beefstew On desktop you can select multiple images at once using shift click. On mobile you can use long click.

**Comment by lastzero** [comment]
Note that we currently have [26 issues actively looking for contributions](https://github.com/photoprism/photoprism/issues?q=is%3Aissue+is%3Aopen+label%3A%22help+wanted%22)  with the "Help Wanted" label. So before anyone asks again, "Is there an update on this?", please check how you can help with those. Thank you! <3

**Comment by dror3go** [comment]
> @beefstew On desktop you can select multiple images at once using shift click. On mobile you can use long click.

I think that what @beefstew meant is not that selecting multiple photos is impossible in mobile devices - but that it's tedious: say I want to add 50 photos to an album, I'll have to long-press one photo, and then tap on the 49 other photos in order to select all of them, while on desktop the shift-click allows a "select from here to here".

Having said that, I feel multi-select on mobile is not part of batch edit feature, although the two are related.

What do you think @graciousgrey?

**Comment by graciousgrey** [comment]
@dror3go 

>  I'll have to long-press one photo, and then tap on the 49 other photos in order to select all of them, while on desktop the shift-click allows a "select from here to here".

"Select from here to here"/multi-select is already possible on mobile, you don't need to tap 49 photos.

1. Do long click on one photo (this selects the photo and starts selection mode)
2. Scroll down and do long click on another photo --> This will select all photos in between

**Comment by rambip** [comment]
What I am missing right now is adding a tag to all selected pictures: it would be so nice !

**Comment by TuRDMaN** [comment]
> What I am missing right now is adding a tag to all selected pictures: it would be so nice !

Yes, this feature is a necessity, in my opinion. Not sure if I can bring myself to import all my photos without some sort of batch label/edit functionality.

**Comment by bigntallmike** [comment]
There are some situations where all of a given location are wrong -- for instance, I've never been to a certain country, but a number of my photos are incorrectly tagged as being taken there. I should be able to look up specific labels or locations, and correct them wholesale -- eg click "London, UK" and delete from *every* photo tagged there.

**Comment by graciousgrey** [comment]
> for instance, I've never been to a certain country, but a number of my photos are incorrectly tagged as being taken there.

Those might be location estimates, you can turn them off in settings (https://docs.photoprism.app/user-guide/settings/library/#estimates).

**Comment by dill0wn** [comment]
I just want to say: Great Product; You now have one more Patron; +1 to this proposed improvement

**Comment by arthuragone** [comment]
Very much looking to this possibility

**Comment by nathcoad** [comment]
Bulk editing labels would be nice too. 

For example if incorrectly categorised with a particular label, it would be nice to view images tagged with that label (already possible) and click an X or something to remove the label from that image.

I'd like to support this feature financially but I see that you're asking people not to use IssueHunt - what is the best way to donate funds specifically towards the implementation of this feature?

**Comment by graciousgrey** [comment]
@nathcoad See https://github.com/photoprism/photoprism/issues/271#issuecomment-1078233571

We kindly ask everyone to not use Github Issues for general questions and opinions. Feel free to use Github Discussions instead. Thank yo very much :heart:

**Comment by maxime1992** [comment]
I really needed this feature to set the missing GPS position on a lot of pictures that were taken at the same place.

I started to do it by hand but then realised it'd probably take me less time to write a script that I could reuse maybe on some other pictures as well than actually doing it one by one on that batch.

**It's clearly hacky**, I didn't want to use something more advanced like Puppeteer, it's a one of and there's this issue so proper support should land at some point. But if that can be of any use for anybody else meanwhile, here's the script I wrote:

```
const bulkEditPosition = async ({ latitude, longitude }) => {
  const elLatitude = document.querySelector('input[aria-label="Latitude"]');
  const elLongitude = document.querySelector('input[aria-label="Longitude"]');

  elLongitude.value = longitude;
  elLongitude.dispatchEvent(new Event('change'));
  elLongitude.dispatchEvent(new Event('keydown'));
  elLongitude.dispatchEvent(new Event('keyup'));
  elLongitude.dispatchEvent(new Event('input'));

  elLatitude.value = latitude;
  elLatitude.dispatchEvent(new Event('change'));
  elLatitude.dispatchEvent(new Event('keydown'));
  elLatitude.dispatchEvent(new Event('keyup'));
  elLatitude.dispatchEvent(new Event('input'));

  await new Promise((res) => {
    setTimeout(() => res(), 200);
  });

  document.querySelector('.action-apply').click();

  await new Promise((res) => {
    setTimeout(() => res(), 2000);
  });

  if (!document.querySelector('button.action-next.v-btn--disabled')) {
    document.querySelector('button.action-next').click();

    await new Promise((res) => {
      setTimeout(() => res(), 3000);
    });

    bulkEditPosition({ latitude, longitude });
  }
};

await bulkEditPosition({
  latitude: xx.xxxxxxxxx, // put your latitude and longitude taken from google maps or whatever here
  longitude: xx.xxxxxxxxx,
});
```

To have this working, select all the pics you want to edit with the multi selection feature and click the edit button. Once you're the page with the GPS coordinates for the first pic, open the dev tool console and launch the script. Enjoy a :coffee: while it's doing its thing :robot:.

PS: For anyone wondering what's the simplest way to grab the correct latitude and longitude, open up google maps, put a point manually where the photo was taken till you get the red point on the map. Then right click onto it and you'll get it there:

![image](https://github.com/photoprism/photoprism/assets/4950209/8697d845-8453-4d53-9dad-7750e638fad1)

**Comment by aaronrudkin** [comment]
I just wanted to flag one issue: right now, it's possible to bulk select photos and click edit in a way that makes it _appear_ as though you are bulk editing, but when you press apply, it only applies to one photo. So, for instance, if you select 100 photos without location data all at the same location, then select a location, and apply -- only the last photo you selected will be edited.

(I don't think the developers need to do anything about this; any effort spent implementing a fix is better spent working on the goal of actual batch editing, and the developer has made clear it's a high priority item and they're limited by their current amount of support/resources [and again, thanks for your hard work], just thought the observation might help any other user who isn't sure what happened when they thought they had bulk edited and it didn't show up properly)

**Comment by lastzero** [comment]
> right now, it's possible to bulk select photos and click edit in a way that makes it appear as though you are bulk editing

This allows you to go through the metadata of each picture without having to return to the search results every time. If we can provide this functionality in a way that is clearer and still avoids this, please let us know! :)

Batch editing is an important use case (which is why it has a high priority on our roadmap), but sometimes you still want to edit individual pictures, e.g. to add names to faces (it is unlikely that all pictures show the same people in the same order...).

**Comment by TokkCorp** [comment]
@maxime1992 This is a really great function, thank you! My knowledge of Javascript is very limited, could you give me a pointer what I would have to do to set the date like this? 
If I do the same the values are not saved. 

Edit: I managed it, like this you can set location, timezone and image date

````
const setVueValue= async({querySelector, value}) => {
	await document.querySelector('.p-form-photo-details-meta').__vue__.$data.inputs.find(element => element.$el.className.includes('input-' + querySelector)).setValue(value);

    await new Promise((res) => {
      setTimeout(() => res(), 200);
    });	
};

const setFieldValue= async({querySelector, value}) => {
    var field = document.querySelector('input[aria-label="'+querySelector+'"]');

    field.value = value;
    field.dispatchEvent(new Event('change'));
    field.dispatchEvent(new Event('keydown'));
    field.dispatchEvent(new Event('keyup'));
    field.dispatchEvent(new Event('input'));

    await new Promise((res) => {
        setTimeout(() => res(), 200);
      });
};


const bulkEditPhotos = async ({ latitude, longitude, day, month, year, timezone }) => {

	await setVueValue({ querySelector: 'year', value: year });
	await setVueValue({ querySelector: 'month', value: month });
	await setVueValue({ querySelector: 'day', value: day });
    await setVueValue({ querySelector: 'timezone', value: timezone });
    await setFieldValue({ querySelector: 'Breitengrad', value: latitude });
    await setFieldValue({ querySelector: 'Längengrad', value: longitude });
  
    document.querySelector('.action-apply').click();
  
    await new Promise((res) => {
      setTimeout(() => res(), 2000);
    });
  
    if (!document.querySelector('button.action-next.v-btn--disabled')) {
      document.querySelector('button.action-next').click();
  
      await new Promise((res) => {
        setTimeout(() => res(), 3000);
      });
  
      bulkEditPhotos({ latitude, longitude, day, month, year, timezone });
    }
  };
  
  await bulkEditPhotos({
    latitude: 50.097282, 
    longitude: 12.356514,
    day: 20,
    month: 11,
    year: 2022,
    timezone: "Europe/Prague"
  });
````

**Comment by Galacticminingoperation** [comment]
This is a serious issue with photoprism. It encourages me to use the software completely different from how it’s supposed to used. Instead of adding labels etc, I’m thinking about adding photos to separate albums simply because I can do it in batch, which is not preferable and is bound to cause a massive amount of clutter. I have 30,000 photos and I feel like I chose the wrong software because labeling them would take orders of magnitude longer than it should. 

I know this is a free and kinda open source project and I have no place saying how this should be run, but “priority” seems like an understatement to me. This should be a “extra-priority.”

**Comment by loulou91** [comment]
@maxime1992 in https://github.com/photoprism/photoprism/issues/271#issuecomment-1567469631 you said "I really needed this feature to set the missing GPS position".
As a workaround you can still use an old freeware as Picasa 3.9 (on PC) which still works well and is very easy and efficient to use.
In addition, if you have a gpx file corresponding to your travel, you can user GeoSetter to automatically feed geo position.

**Comment by pnwbob** [comment]
Please make this happen. Would be very nice to have

**Comment by scarolan** [comment]
I also have a ton of photos that were created in the pre-EXIF days and scanned.  Would love batch edit so I can at least get them in the right year for sorting, etc.

**Comment by Mannshoch** [comment]
I have a lot of unrecognized Faces. I which to batch  add to a Person

**Comment by anst84** [comment]
What is the status of this development? This was posted over three years ago, is high-prio in the roadmap/backlog, so why is it taking so long time to get out? Thanks in advance.

**Comment by rambip** [comment]
> What is the status of this development? This was posted over three years ago, is high-prio in the roadmap/backlog, so why is it taking so long time to get out? Thanks in advance.

I also would like to have some idea of the status of this feature

**Comment by stephenchew** [comment]
While waiting indefatigably for the great batch editing feature, I came across this issue and got really inspired by the "workaround" suggested by @maxime1992 and @TokkCorp. Hence, I made a script for myself, and hopefully could help others who are longing for this feature, before the official is released.

You can find the instruction and script at: -
https://gist.github.com/stephenchew/b73ecc75b77a84a92fa350048d5ca84f

**Comment by karthikiyengar** [comment]
@stephenchew Wow, thanks a lot. Wondering if we can turn this into a greasemonkey/tampermonkey script  with some more effort. Haven't looked into the details yet, but it'd be amazing to have a console-less experience by tweaking the foundations here.

**Comment by b-onc** [comment]
@stephenchew thanks a lot! can you add day/month/year fields to the gist too? And also, do we run the `run.js` script in devtools console (copy/paste) too?

**Comment by stephenchew** [comment]
> @stephenchew Wow, thanks a lot. Wondering if we can turn this into a greasemonkey/tampermonkey script with some more effort. Haven't looked into the details yet, but it'd be amazing to have a console-less experience by tweaking the foundations here.

Good idea 😃💡 I must confess that the last I installed GreaseMonkey was a decade ago, I'm not entirely familiar with the building blocks. I shall check it out when I have time.

> @stephenchew thanks a lot! can you add day/month/year fields to the gist too? And also, do we run the `run.js` script in devtools console (copy/paste) too?

I'll look into the "date" field. I hope it's straightforward.
Yes, just copy the content of `run.js` and paste it in DevTools console. You can either comment out / remove the attributes (or the `content` field of the attribute) that you do not want to update.

**Comment by stephenchew** [comment]
@b-onc The script is now updated to modify "day", "month" and "year". It comes with validation so you shouldn't be able to enter invalid values 🤞🏻

**Comment by michaill** [comment]
Is this feature going to be implemented? or it is just in bottom of the backlog. Feature requested by so many people (UX already solved here - save all button) is missing for three years. Workaround is not working for all attributes and is little bit hacky.

**Comment by maxime1992** [comment]
> Is this feature going to be implemented? or it is just in bottom of the backlog

You can see the board [here](https://github.com/orgs/photoprism/projects/5)

> is missing for three years

It's an open source project, feel free to contribute or see if you can put a bounty on the issue so that maybe more people can be hired to work on the project full time

**Comment by Blendan1** [comment]
Hello,

I'm currently working on an inbetween solution, a chrome extention that allowes you to bulk edit. 
It's based on the gist from @stephenchew 

You can find it here: https://github.com/Blendan1/PhotoPrismBulkEditor

Hope that helps some untill the offical way comes

**Comment by filipepatricio** [comment]
> > Is this feature going to be implemented? or it is just in bottom of the backlog
> 
> You can see the board [here](https://github.com/orgs/photoprism/projects/5)
> 
> > is missing for three years
> 
> It's an open source project, feel free to contribute or see if you can put a bounty on the issue so that maybe more people can be hired to work on the project full time

How can I put a bounty on this issue?
Thank you very much

**Comment by lastzero** [comment]
We will get to this. However, as communicated, we will work on OpenID Connect and 2-factor authentication first, as many users are waiting for that as well.

**Comment by Moosbueffel** [comment]
As mentioned a lot of time, it's an basic feature to edit metadata for a lot of (new) pictures. So I will share my toughts for perhaps a dayjob to implement- but don't kill me, if it's more work, I'm not involved ;).

I assume, you have 2-n pictures selected for editing. The UX with the edit dialog is a disaster for me. Why? In the first tab (Details) I have to press button save, in the other tab's it's save autmatically. And if I forget saving and got on to next picture my work is done. Why not save at the end of my procedure like I do in "all" other apps. I can save every time, or at the end with Ctrl+s (or respective button). Or your UI work always with immediatly savings.

So - in this context I think you plan to revise this part of software. And as a interim solution **how about to add a "save&copy" button**? With this you do the save as now and additional copy all entered fields to the next picture and switch to this picture, This has the effort of comfortable editing. I have only one click (save and goToNextPic) and save the time to input all text again and again. Also this solution let you adjust/change some things in the next picture. If you don't want changes, just press "save&copy"

It's difficult to describe and my english is not really well, but I hope you understand, what I'm thinking. Feel free to ask back. And perhaps you can discuss it in your team, have praktikant, diplomant, some of this forum or any other person who can do this ;). If not it is perhals a postIt for the time it will be realized.

Many thanks for your great work and ideas! Over all - it's christmas time and time to support you :)

**Comment by matchstick** [comment]
Just found photoprism and am super excited. I have TBs of photos and indexing is going great. This bug though is the only thing that is keeping me from being a subscriber. Is there anything we can do to help resolve batch editing of meta data? Is it funds or just not the highest in the priority queue?

**Comment by boecko** [comment]
> @stephenchew Wow, thanks a lot. Wondering if we can turn this into a greasemonkey/tampermonkey script with some more effort. Haven't looked into the details yet, but it'd be amazing to have a console-less experience by tweaking the foundations here.

i've created a greasemonkey-script out of @stephenchew s gist. 
 see https://gist.github.com/boecko/e2d0effe7c61976c22e6bc0a8ee645c7#file-photoprismbulkeditor-user-js

installlink: https://gist.github.com/boecko/e2d0effe7c61976c22e6bc0a8ee645c7/raw/photoprismbulkeditor.user.js

You can still run `runBulk` in the console, but it i've extended the edit-view of photoprismn.
![bulkchange](https://github.com/photoprism/photoprism/assets/129901/7a6588cb-0d43-4662-b327-2d233894788d)

**Comment by isleshocky77** [comment]
@boecko 
> i've created a greasemonkey-script out of @stephenchew s gist. see https://gist.github.com/boecko/e2d0effe7c61976c22e6bc0a8ee645c7

If you change the filename in the gist to `*.user.js` it makes it installable which is pretty useful. A couple year old [example](https://gist.github.com/isleshocky77/467c9717cda473c68664e460d8c3fd07) of something I made for people to install themselves. You can even add an `updateUrl` for people to check for updates.

**Comment by boecko** [comment]
> @boecko
> 
> > i've created a greasemonkey-script out of @stephenchew s gist. see https://gist.github.com/boecko/e2d0effe7c61976c22e6bc0a8ee645c7
> 
> If you change the filename in the gist to `*.user.js` it makes it installable which is pretty useful. A couple year old [example](https://gist.github.com/isleshocky77/467c9717cda473c68664e460d8c3fd07) of something I made for people to install themselves. You can even add an `updateUrl` for people to check for updates.

done

**Comment by Moosbueffel** [comment]
@boecko Thanks for the script and soory for the help-question, but how to use it?
I found a way to install the script in firefox with the greasemoney addon. There it is marked as installed and active. And now??

As I understand right, I should see the bulkeditor In PhotoPrism, but there is no change. Is there any additonal to do? And how does it work? Can I than mark some pics and edit all with the same field content?

Thanks

**Comment by boecko** [comment]
> @boecko Thanks for the script and soory for the help-question, but how to use it? I found a way to install the script in firefox with the greasemoney addon. There it is marked as installed and active. And now??
> 
> As I understand right, I should see the bulkeditor In PhotoPrism, but there is no change. Is there any additonal to do? And how does it work? Can I than mark some pics and edit all with the same field content?
> 
> Thanks

Maybe you have to adjust the @ match lines in case you don't use https
![image](https://github.com/photoprism/photoprism/assets/129901/fcb0dfd0-d9e0-441b-8547-dda00a784b8a)

**Comment by Moosbueffel** [comment]
@boecko Thanks, now it works. If you can write a short description ;).

- which photos are effected?
- In the source as saw a right-mouse-click?
- is bulkchange valid for all fields, or only the title?

e.g. if I are in browse mode with filter 2022 than a change of the title an pressing "bulkchange", than all photos of 2022-  will be changed to the new title? (and similar in other filter, kategorie,... )

**Comment by maxime1992** [comment]
Discussions for hacks start to generate quite a lot of messages. I think it's really nice that people mention what they found out to unlock others until we have an official feature for this but wouldn't it be great if people created a repo or a gist where their own hack can be discussed separately? As in drop a message here showcasing your code/extension/whatever and invite people who are interested to join a different thread to discuss it further.

I know that personally I'd prefer to avoid the noise except for questions/news about the implementation of an official feature. 

_(Also, do note that I am no maintainer of this repo so if everyone is happy with the current way the discussion is going feel free to discard me, it's just a personal preference)_

**Comment by boecko** [comment]
@Moosbueffel you can comment on my gist https://gist.github.com/boecko/e2d0effe7c61976c22e6bc0a8ee645c7

**Comment by lastzero** [comment]
You are welcome to send PRs with your own scripts, libraries, tutorials, or config examples to the following repository that we maintain for this purpose: https://github.com/photoprism/photoprism-contrib

**Comment by zane-shus** [comment]
> @Moosbueffel you can comment on my gist https://gist.github.com/boecko/e2d0effe7c61976c22e6bc0a8ee645c7

I am sure this is a silly question, but where do I save this file, i.e. how do i install it?

Also, I cant work it out from the screen shot, will this also bulk Label as well as change detail?

**Comment by dulio** [comment]
I really need to bulk edit pictures

**Comment by bvvl** [comment]
I started using Photoprism a couple of months ago.  Also found editing metadata cumbersome so I wrote a chrome extension to to help me out.  I started with Blendan1's version but wanted something a little more flexible.
Have a look at the following if you like:
https://github.com/bvvl/Bulk-Edit-for-Photoprism/

**Comment by matt-benett** [comment]
I would like to use Photo Prism at work, but missing the ability to natively Batch Tag images is the one feature I cant work around.  Is there a way I can fund this feature request up the priority list?

**Comment by lastzero** [comment]
@matt-benett No worries, we'll get to this as soon as we've released the additional authentication options and much-needed facial recognition improvements (any help with this is much appreciated and will speed things up). See our [public roadmap](https://github.com/orgs/photoprism/projects/5) to find out which issues need help and what's next.

**Comment by meevil** [comment]
> @matt-benett No worries, we'll get to this as soon as we've released the additional authentication options and much-needed facial recognition improvements (any help with this is much appreciated and will speed things up). See our [public roadmap](https://github.com/orgs/photoprism/projects/5) to find out which issues need help and what's next.

This request is quite old, however it's exactly what we would need to use photoprism in our company. As @matt-benett already asekd: Is there a way to support you project to make this feature available soon? Thanks!

**Comment by alelom** [comment]
> I started using Photoprism a couple of months ago. Also found editing metadata cumbersome so I wrote a chrome extension to to help me out. I started with Blendan1's version but wanted something a little more flexible. Have a look at the following if you like: https://github.com/bvvl/Bulk-Edit-for-Photoprism/

Just restating what other say -- batch edit is a very needed feature. 

Adding that the above-mentioned Chrome extension works for me, with the latest Photoprism ([`photoprism_240420-ef5f14bc4-linux-amd64`](https://github.com/photoprism/photoprism/releases/download/240420-ef5f14bc4/photoprism_240420-ef5f14bc4-linux-amd64.tar.gz) running on Ubuntu Server 24) and Chrome (`124.0.6367.201 (Official Build) (64-bit)` on Windows). 

Very grateful for it @bvvl ! 🚀  ❤️

**Comment by staticnom** [comment]
found photoprism with all the ai hype and I really want to ditch google photos. batch edit is a requirement, I can't believe this request is open since 2020!?

**Comment by rccm8** [comment]
Looking forward to batch editing. I think it'll be a great feature for Photoprism

**Comment by balki** [comment]
+1 for batch editing. I prefer manual labels. Automatic labels are mostly incorrect and not really useful even when it is correct.

**Comment by friendlyantz** [comment]
+1 for bulk edit/ gps assignment.

**Comment by CarcajadaArtificial** [comment]
+1 for batch editing, come on guys

**Comment by graciousgrey** [comment]
Please give this issue a 👍️ instead of leaving +1 comments to vote for it, as our team and all issue subscribers receive an email notification from GitHub whenever a new comment is added. So these should only be used for sharing important information. Thank you very much!

**Here is a draft and a brief description of what is planned:**

![batch-edit-draft](https://github.com/user-attachments/assets/5576ba79-3f57-4297-973a-036ebda213c1)

- If several images are selected, the batch edit form shown above may be opened via the context menu.
- On the left side you can see which images are selected. From there, you can open them in full-screen mode or remove them from the selection.
- Initially, there will probably be an option for each data field to either overwrite it or retain the current value. Later it would also be conceivable to offer an “Append” or “Prepend” option for certain fields (e.g. text fields such as description, copyright, notes or license). A “delete” option for deleting existing values would also be considerable.
- In the lower area, it should be possible to add or remove labels and keywords.

**Ideas for further improvements after the initial implementation:**

- Option to save the filled form as a template

- Option to set coordinates via a map or address

- Option to shift the date/time by a certain number of days, hours seconds


Note that in order to avoid duplicate work and merge conflicts, we can only start implementing this once the upgrade of our frontend component framework is complete:

- https://github.com/photoprism/photoprism/issues/3168

**Comment by cko-skverlag** [comment]
Looking really forward to this feature - what is your expected publishing date?

**Comment by MyWebmasterD** [comment]
+1

**Comment by freakhauser** [comment]
+1

**Comment by cko-skverlag** [comment]
@omerdduran could you help us with a potential publishing date for this feature? thanks

**Comment by omerdduran** [comment]
Currently this feature is working but there are some optimization issues. Also, we are currently working on the ollama updates and face recognition updates. Batch edit feature will not be available in the next release but hopefully in the one after that.

**Comment by lastzero** [comment]
I'm happy to report that you can now test batch editing with the [development preview builds](https://docs.photoprism.app/release-notes/#development-preview) and/or on our [public demo](https://demo.photoprism.app/) (see the screenshot below). More information and documentation will follow. Thank you all for your support and patience while we got this ready for you! 🤗

<img width="1400" height="950" alt="Image" src="https://github.com/user-attachments/assets/411bac50-63ac-415c-91e9-e135ab949fe3" />

**Comment by lastzero** [comment]
I'm happy to announce that Batch Edit is now available as part of our latest stable release:
- https://github.com/photoprism/photoprism/releases/tag/251130-b3068414c
- https://docs.photoprism.app/getting-started/updates/
